boa/src/syntax/lexer/cursor.rs

```diff
@@ -0,0 +1,243 @@
+            peeked: None,
+            pos: Position::new(1, 1),
+        }
+    }
+
+    /// Peeks the next character.
+    #[inline]
+    pub(super) fn peek(&mut self) -> Result<Option<char>, Error> {
+        let iter = &mut self.iter;
+        if let Some(v) = self.peeked {
+            Ok(v)
+        } else {
+            let val = iter.next().transpose()?;
+            self.peeked = Some(val);
+            Ok(val)
+        }
+    }
+
+    /// Compares the character passed in to the next character, if they match true is returned and the buffer is incremented
+    #[inline]
+    pub(super) fn next_is(&mut self, peek: char) -> io::Result<bool> {
+        Ok(match self.peek()? {
+            Some(next) if next == peek => {
+                let _ = self.peeked.take();
+                true
+            }
+            _ => false,
+        })
+    }
+
+    /// Applies the predicate to the next character and returns the result.
+    /// Returns false if there is no next character.
+    ///
+    /// The buffer is not incremented.
+    #[inline]
+    pub(super) fn next_is_pred<F>(&mut self, pred: &F) -> io::Result<bool>
+    where
+        F: Fn(char) -> bool,
+    {
+        Ok(if let Some(peek) = self.peek()? {
+            pred(peek)
+        } else {
+            false
+        })
+    }
+
+    /// Fills the buffer with all characters until the stop character is found.
+    ///
+    /// Note: It will not add the stop character to the buffer.
+    pub(super) fn take_until(&mut self, stop: char, buf: &mut String) -> io::Result<()> {
+        loop {
+            if self.next_is(stop)? {
+                return Ok(());
+            } else if let Some(ch) = self.next()? {
+                buf.push(ch);
+            } else {
+                return Err(io::Error::new(
+                    ErrorKind::UnexpectedEof,
+                    format!("Unexpected end of file when looking for character {}", stop),
+                ));
+            }
+        }
+    }
+
+    /// Fills the buffer with characters until the first character (x) for which the predicate (pred) is false
+    /// (or the next character is none).
+    ///
+    /// Note that all characters up until x are added to the buffer including the character right before.
+    pub(super) fn take_until_pred<F>(&mut self, buf: &mut String, pred: &F) -> io::Result<()>
+    where
+        F: Fn(char) -> bool,
+    {
+        loop {
+            if !self.next_is_pred(pred)? {
+                return Ok(());
+            } else if let Some(ch) = self.next()? {
+                buf.push(ch);
+            } else {
+                // next_is_pred will return false if the next value is None so the None case should already be handled.
+                unreachable!();
+            }
+        }
+    }
+
+    /// It will fill the buffer with checked ASCII bytes.
+    pub(super) fn fill_bytes(&mut self, buf: &mut [u8]) -> io::Result<()> {
+        unimplemented!("Lexer::cursor::fill_bytes {:?}", buf)
+    }
```

Comment:
Do we need this?

Reply:
I don't think so, its from the string_literal code and I assumed you were using it for a specific reason so I left it

Reply:
Are we still using it in string literals, or did you find a better way to do it?

Reply:
> Are we still using it in string literals, or did you find a better way to do it?

I didn't really change the string literal stuff so its still there but equally all the tests pass so it can't be doing anything major at the moment - will need to look back into the source code and see

Reply:
Actually I think we might need it - the only usage seems to require the raw u8 bytes rather than a rust char (and so therefore the regular .next() method won't work).